test/e2e/upgrade_hypershift_operator_test.go

```diff
@@ -0,0 +1,226 @@
+
+				err = mgmtClient.List(ctx, machineDeployments, crclient.InNamespace(hcpNameSpace))
+
+				g.Expect(err).ToNot(gomega.HaveOccurred(),
+					"Listing MachineDeployments in namespace %s shouldn't return errors", hcpNameSpace)
+				g.Expect(machineDeployments.Items).ToNot(gomega.BeEmpty(),
+					"Should find MachineDeployments in namespace %s", hcpNameSpace)
+				g.Expect(len(machineDeployments.Items)).To(gomega.BeEquivalentTo(len(nodepools.Items)),
+					"Number of MachineDeployments and NodePools should match")
+
+				machineDeploymentMap = make(map[string]*v1beta1.MachineDeployment)
+				t.Logf("Found %d MachineDeployments", len(machineDeployments.Items))
+				for i := range machineDeployments.Items {
+					t.Logf("Found MachineDeployment %s", machineDeployments.Items[i].Name)
+					machineDeploymentMap[machineDeployments.Items[i].Name] = &machineDeployments.Items[i]
+				}
+
+			})).To(gomega.BeTrue(), "Calculating HyperShift Operator upgrade invariants should succeed")
+
+			g.Expect(t.Run("Upgrade HyperShift Operator", func(t *testing.T) {
+				t.Logf("Upgrading HyperShift Operator to image %s", hyperShiftOperatorLatestImage)
+				globalOpts.HyperShiftOperatorLatestImage = hyperShiftOperatorLatestImage
+				// Note that we're replacing previous HO installation cleanup function with the new one
+				cleanup, err = e2eutil.InstallHyperShiftOperator(ctx, globalOpts)
+				if err != nil {
+					t.Fatal("Failed to upgrade HyperShift Operator")
+				}
+			})).To(gomega.BeTrue(), "Upgrade HyperShift Operator should succeed")
+
+			g.Expect(t.Run("Verify upgrade invariants", func(t *testing.T) {
+				t.Log("Verifying upgrade invariants")
+				namespace := hostedCluster.Namespace
+				g := gomega.NewWithT(t)
+
+				postUpgradeNodePools := &hyperv1.NodePoolList{}
+				err = mgmtClient.List(ctx, postUpgradeNodePools, crclient.InNamespace(namespace))
+				g.Expect(err).ToNot(gomega.HaveOccurred(), "Listing NodePools in namespace %s shouldn't return errors",
+					namespace)
+
+				g.Expect(postUpgradeNodePools.Items).ToNot(gomega.BeEmpty(), "Should find NodePools in namespace %s",
+					namespace)
+				g.Expect(len(postUpgradeNodePools.Items)).To(gomega.BeEquivalentTo(len(nodePoolsMap)),
+					"Number of pre-ugrade and post-upgrade NodePools should match")
+
+				for _, nodePool := range postUpgradeNodePools.Items {
+					t.Logf("Verifying NodePool %s", nodePool.Name)
+					preUpgradeNodePool, ok := nodePoolsMap[nodePool.Name]
+					g.Expect(ok).To(gomega.BeTrue(), "Should have a pre-upgrade nodepool with the name %s",
+						nodePool.Name)
+
+					t.Logf("Generation: Want %d, Got %d", preUpgradeNodePool.Generation, nodePool.Generation)
+					t.Logf("CurrentConfig: Want %s, Got %s",
+						preUpgradeNodePool.Annotations[nodePoolAnnotationCurrentConfig],
+						nodePool.Annotations[nodePoolAnnotationCurrentConfig])
+					t.Logf("CurrentConfigVersion: Want %s, Got %s",
```

Comment:
we are checking those values with `g.Expect` down, right? why do we need these logs? I think it will make it harder to find the real error if there is one.

Reply:
I think it's crucial to have visibility over the test invariants in the case of success as well, to detect any error in the logic or any bug in the test. 

Here's an example output for the case of a failure, do you find it harder to find the real error?

```
=== RUN   TestUpgradeHyperShiftOperator/Main/Verify_upgrade_invariants
    upgrade_hypershift_operator_test.go:162: Verifying upgrade invariants
    upgrade_hypershift_operator_test.go:177: Verifying NodePool example-gddm2-us-east-1a
    upgrade_hypershift_operator_test.go:182: Generation: Want 1, Got 2
    upgrade_hypershift_operator_test.go:183: CurrentConfig: Want 00d3f987, Got 00d3f987
    upgrade_hypershift_operator_test.go:186: CurrentConfigVersion: Want f287df15, Got f287df15
    upgrade_hypershift_operator_test.go:191: 
        Pre-upgrade and post-upgrade NodePool generations should match
        Expected
            <int64>: 1
        not to equal
            <int64>: 2
=== NAME  TestUpgradeHyperShiftOperator/Main
```

Reply:
> to detect any error in the logic or any bug in the test.

you might have a bug in the logs only, since you got duplicate code now. 

> Here's an example output for the case of a failure, do you find it harder to find the real error?

right, you can see the error at the bottom, but I got distracted reading the logs at beginning first. 
Specifically this `CurrentConfig: Want 00d3f987, Got 00d3f987` as it's not easy to see if they match or not immediately :D

again no strong feelings here.

Reply:
Got it, yeah I can see your point. 

Okay will push a version that's just a log of the current state without the want and got and lemme know your thoughts.

Reply:
Done. Now it'd look like this


```
=== RUN   TestUpgradeHyperShiftOperator/Main/Verify_upgrade_invariants
    upgrade_hypershift_operator_test.go:162: Verifying upgrade invariants
    upgrade_hypershift_operator_test.go:177: Verifying NodePool example-gddm2-us-east-1a
    upgrade_hypershift_operator_test.go:182: Generation:  2
    upgrade_hypershift_operator_test.go:183: CurrentConfig:  00d3f987
    upgrade_hypershift_operator_test.go:186: CurrentConfigVersion: f287df15
    upgrade_hypershift_operator_test.go:191: 
        Pre-upgrade and post-upgrade NodePool generations should match
        Expected
            <int64>: 1
        not to equal
            <int64>: 2
=== NAME  TestUpgradeHyperShiftOperator/Main
```

Thoughts?

Reply:
thanks, looks better. Why does it say  "Expected <int64>: 1 `not to equal` <int64>: 2" though? shouldn't it be saying `to equal`

Reply:
hahaha, the only way for me to get the test to fail is to change the condition temporary to "NotEqual" instead of "Equal" 😄 , so the actual failure I got on my machine was  `Expected 1 not to equal 1` .

But when I copied it here, I changed the 2nd value to 2 to match the real error that would show in the test, but I forgot to change the `not to equal` to `equal` 😅 

So yeah, it'd actually be

```
        Pre-upgrade and post-upgrade NodePool generations should match
        Expected
            <int64>: 1
         to equal
            <int64>: 2
```

Reply:
fair enough XD